File: aop/aop-annotation-processor/src/main/java/ru/tinkoff/kora/aop/annotation/processor/AopProcessor.java
package ru.tinkoff.kora.aop.annotation.processor;

import com.squareup.javapoet.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import ru.tinkoff.kora.annotation.processor.common.*;
import ru.tinkoff.kora.common.Component;
import ru.tinkoff.kora.common.annotation.Generated;

import javax.lang.model.element.Modifier;
import javax.lang.model.element.TypeElement;
import javax.lang.model.type.TypeKind;
import javax.lang.model.util.Elements;
import javax.lang.model.util.Types;
import java.util.*;

public class AopProcessor {
    private static final Logger log = LoggerFactory.getLogger(AopProcessor.class);

    private final List<KoraAspect> aspects;
    private final Types types;
    private final Elements elements;

    public AopProcessor(Types types, Elements elements, List<KoraAspect> aspects) {
        this.aspects = aspects;
        this.types = types;
        this.elements = elements;
    }

    private static class TypeFieldFactory implements KoraAspect.FieldFactory {
        private final Types types;
        private final Set<String> fieldNames = new HashSet<>();
        private final Map<ConstructorParamKey, String> constructorParams = new LinkedHashMap<>();
        private final Map<ConstructorInitializedParamKey, String> constructorInitializedParams = new LinkedHashMap<>();

        private record ConstructorParamKey(TypeName type, List<AnnotationSpec> annotations, Types types) {}

        private record ConstructorInitializedParamKey(TypeName type, CodeBlock initializer, Types types) {}

        private TypeFieldFactory(Types types) {
            this.types = types;
        }

        public void addFields(TypeSpec.Builder typeBuilder) {
            for (var entry : this.constructorParams.entrySet()) {
                var name = entry.getValue();
                var fd = entry.getKey();
                typeBuilder.addField(FieldSpec.builder(fd.type(), name, Modifier.PRIVATE, Modifier.FINAL)
                    .build());
            }
            for (var entry : this.constructorInitializedParams.entrySet()) {
                var name = entry.getValue();
                var fd = entry.getKey();
                typeBuilder.addField(FieldSpec.builder(fd.type(), name, Modifier.PRIVATE, Modifier.FINAL)
                    .build());
            }
        }

        @Override
        public String constructorParam(TypeName type, List<AnnotationSpec> annotations) {
            return this.constructorParams.computeIfAbsent(new ConstructorParamKey(type, annotations, this.types), this::computeFieldName);
        }

        @Override
        public String constructorInitialized(TypeName type, CodeBlock initializer) {
            return this.constructorInitializedParams.computeIfAbsent(new ConstructorInitializedParamKey(type, initializer, this.types), this::computeFieldName);
        }

        public void enrichConstructor(MethodSpec.Builder constructorBuilder) {
            for (var entry : this.constructorParams.entrySet()) {
                var name = entry.getValue();
                var fd = entry.getKey();
                constructorBuilder.addParameter(ParameterSpec.builder(fd.type(), name)
                    .addAnnotations(fd.annotations())
                    .build());
                constructorBuilder.addCode("this.$L = $L;\n", name, name);
            }
            for (var entry : this.constructorInitializedParams.entrySet()) {
                var name = entry.getValue();
                var fd = entry.getKey();
                constructorBuilder.addCode("this.$L = $L;\n", name, fd.initializer);
            }
        }

        private String computeFieldName(ConstructorParamKey key) {
            var qualifiedType = key.type().toString();
            if (qualifiedType.indexOf('<') > 0) {
                qualifiedType = qualifiedType.substring(0, qualifiedType.indexOf('<'));
            }

            var dotIndex = qualifiedType.lastIndexOf('.');
            var shortName = dotIndex < 0
                ? CommonUtils.decapitalize(qualifiedType)
                : CommonUtils.decapitalize(qualifiedType.substring(dotIndex + 1));
            for (int i = 1; i < Integer.MAX_VALUE; i++) {
                var name = shortName + i;
                if (this.fieldNames.add(name)) {
                    return name;
                }
            }
            // never gonna happen

            throw new IllegalStateException("Can't compute name for " + key);
        }

        private String computeFieldName(ConstructorInitializedParamKey key) {
            var qualifiedType = key.type().toString();
            if (qualifiedType.indexOf('<') > 0) {
                qualifiedType = qualifiedType.substring(0, qualifiedType.indexOf('<'));
            }

            var dotIndex = qualifiedType.lastIndexOf('.');
            var shortName = dotIndex < 0
                ? CommonUtils.decapitalize(qualifiedType)
                : CommonUtils.decapitalize(qualifiedType.substring(dotIndex + 1));
            for (int i = 1; i < Integer.MAX_VALUE; i++) {
                var name = shortName + i;
                if (this.fieldNames.add(name)) {
                    return name;
                }
            }
            // never gonna happen
            return null;
        }
    }


    public TypeSpec applyAspects(TypeElement typeElement) {
        var constructor = AopUtils.findAopConstructor(typeElement);
        if (constructor == null) {
            throw new ProcessingErrorException("Class has no aop suitable constructor", typeElement);
        }
        var typeLevelAspects = new ArrayList<KoraAspect>();
        for (var am : typeElement.getAnnotationMirrors()) {
            for (var aspect : this.aspects) {
                if (aspect.getSupportedAnnotationTypes().contains(am.getAnnotationType().toString())) {
                    if (!typeLevelAspects.contains(aspect)) {
                        typeLevelAspects.add(aspect);
                    }
                }
            }
        }
        log.trace("Type level aspects for {}: {}", typeElement, typeLevelAspects);

        var typeFieldFactory = new TypeFieldFactory(this.types);
        var aopContext = new KoraAspect.AspectContext(typeFieldFactory);

        var typeBuilder = TypeSpec.classBuilder(AopUtils.aopProxyName(typeElement))
            .superclass(typeElement.asType())
            .addModifiers(Modifier.PUBLIC, Modifier.FINAL)
            .addAnnotation(CommonClassNames.aopProxy);

        var tag = TagUtils.parseTagValue(typeElement);
        if (tag != null && !tag.isEmpty()) {
            typeBuilder.addAnnotation(TagUtils.makeAnnotationSpec(tag));
        }
        if (AnnotationUtils.isAnnotationPresent(typeElement, CommonClassNames.root)) {
            typeBuilder.addAnnotation(CommonClassNames.root);
        }

        var appliedProcessors = new LinkedHashSet<String>();
        appliedProcessors.add(AopAnnotationProcessor.class.getCanonicalName());

        var typeMethods = CommonUtils.findMethods(typeElement, m -> !m.contains(Modifier.STATIC) && (m.contains(Modifier.PROTECTED) || m.contains(Modifier.PUBLIC)));
        for (var typeMethod : typeMethods) {
            var methodLevelTypeAspects = new ArrayList<>(typeLevelAspects);
            var methodLevelAspects = new ArrayList<KoraAspect>();
            var methodParameterLevelAspects = new ArrayList<KoraAspect>();
            for (var am : typeMethod.getAnnotationMirrors()) {
                for (var aspect : this.aspects) {
                    if (aspect.getSupportedAnnotationTypes().contains(am.getAnnotationType().toString())) {
                        if (!methodLevelAspects.contains(aspect)) {
                            methodLevelAspects.add(aspect);
                        }
                        methodLevelTypeAspects.remove(aspect);
                    }
                }
            }
            for (var parameter : typeMethod.getParameters()) {
                for (var am : parameter.getAnnotationMirrors()) {
                    for (var aspect : this.aspects) {
                        if (aspect.getSupportedAnnotationTypes().contains(am.getAnnotationType().toString())) {
                            if (!methodParameterLevelAspects.contains(aspect) && !methodLevelAspects.contains(aspect)) {
                                methodParameterLevelAspects.add(aspect);
                            }
                            methodLevelTypeAspects.remove(aspect);
                        }
                    }
                }
            }

            if (methodLevelTypeAspects.isEmpty() && methodLevelAspects.isEmpty() && methodParameterLevelAspects.isEmpty()) {
                continue;
            }
            log.trace("Method level aspects for {}#{}: {}", typeElement, typeMethod.getSimpleName(), methodLevelAspects);
            var aspectsToApply = new ArrayList<>(methodLevelTypeAspects);
            aspectsToApply.addAll(methodLevelAspects);
            aspectsToApply.addAll(methodParameterLevelAspects);
            var superCall = "super." + typeMethod.getSimpleName();
            var overridenMethod = MethodSpec.overriding(typeMethod);
            Collections.reverse(aspectsToApply);

            for (var aspect : aspectsToApply) {
                var result = aspect.apply(typeMethod, superCall, aopContext);
                if (result instanceof KoraAspect.ApplyResult.Noop) {
                    continue;
                }

                var methodBody = (KoraAspect.ApplyResult.MethodBody) result;
                var methodName = "_" + typeMethod.getSimpleName() + "_AopProxy_" + aspect.getClass().getSimpleName();
                superCall = methodName;
                var m = MethodSpec.methodBuilder(methodName)
                    .addModifiers(Modifier.PRIVATE)
                    .addCode(methodBody.codeBlock());
                for (var parameter : typeMethod.getParameters()) {
                    m.addParameter(ParameterSpec.builder(TypeName.get(parameter.asType()), parameter.getSimpleName().toString()).build());
                }
                for (var typeParameter : typeMethod.getTypeParameters()) {
                    m.addTypeVariable(TypeVariableName.get(typeParameter));
                }
                for (var exception : typeMethod.getThrownTypes()) {
                    m.addException(TypeName.get(exception));
                }
                m.returns(TypeName.get(typeMethod.getReturnType()));
                typeBuilder.addMethod(m.build());

                appliedProcessors.add(aspect.getClass().getCanonicalName());
            }
            var b = CodeBlock.builder();
            if (typeMethod.getReturnType().getKind() != TypeKind.VOID) {
                b.add("return ");
            }
            if (superCall.startsWith("super.")) {
                b.add("$L(", superCall);
            } else {
                b.add("this.$L(", superCall);
            }
            for (int i = 0; i < typeMethod.getParameters().size(); i++) {
                if (i > 0) {
                    b.add(", ");
                }
                var parameter = typeMethod.getParameters().get(i);
                b.add("$L", parameter);
            }
            b.add(");\n");
            overridenMethod.addCode(b.build());
            typeBuilder.addMethod(overridenMethod.build());
        }

        var generated = appliedProcessors.stream()
            .map(a -> CodeBlock.of("$S", a))
            .collect(CodeBlock.joining(", ", "{", "}"));

        typeBuilder
            .addAnnotation(AnnotationSpec.builder(Generated.class)
                .addMember("value", generated)
                .build());

        if (typeElement.getAnnotation(Component.class) != null) {
            typeBuilder.addAnnotation(Component.class);
        }

        var constructorBuilder = MethodSpec.constructorBuilder()
            .addModifiers(Modifier.PUBLIC);
        constructorBuilder.addCode("super(");
        for (int i = 0; i < constructor.getParameters().size(); i++) {
            if (i > 0) {
                constructorBuilder.addCode(", ");
            }
            var parameter = constructor.getParameters().get(i);
            constructorBuilder.addCode("$L", parameter);
            var parameterSpec = ParameterSpec.get(parameter);
            var tags = TagUtils.parseTagValue(parameter);

            if (!tags.isEmpty()) {
                parameterSpec = parameterSpec.toBuilder().addAnnotation(TagUtils.makeAnnotationSpec(tags)).build();
            }

            constructorBuilder.addParameter(parameterSpec);
        }
        constructorBuilder.addCode(");\n");
        typeFieldFactory.addFields(typeBuilder);
        typeFieldFactory.enrichConstructor(constructorBuilder);
        typeBuilder.addMethod(constructorBuilder.build());

        return typeBuilder.build();
    }
}

File: aop/aop-symbol-processor/src/main/kotlin/ru/tinkoff/kora/aop/symbol/processor/AopProcessor.kt
package ru.tinkoff.kora.aop.symbol.processor

import com.google.devtools.ksp.isConstructor
import com.google.devtools.ksp.isProtected
import com.google.devtools.ksp.isPublic
import com.google.devtools.ksp.processing.Resolver
import com.google.devtools.ksp.symbol.KSClassDeclaration
import com.google.devtools.ksp.symbol.Modifier
import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ksp.toClassName
import com.squareup.kotlinpoet.ksp.toTypeName
import com.squareup.kotlinpoet.ksp.toTypeVariableName
import ru.tinkoff.kora.ksp.common.*
import ru.tinkoff.kora.ksp.common.AnnotationUtils.isAnnotationPresent
import ru.tinkoff.kora.ksp.common.KspCommonUtils.generated
import ru.tinkoff.kora.ksp.common.KspCommonUtils.resolveToUnderlying
import ru.tinkoff.kora.ksp.common.exception.ProcessingErrorException
import kotlin.reflect.KClass

class AopProcessor(private val aspects: List<KoraAspect>, private val resolver: Resolver) {

    private class TypeFieldFactory(private val resolver: Resolver) : KoraAspect.FieldFactory {
        private val fieldNames: MutableSet<String> = HashSet()
        private val constructorParams: MutableMap<ConstructorParamKey, String> = linkedMapOf()
        private val constructorInitializedParams: MutableMap<ConstructorInitializedParamKey, String> = linkedMapOf()

        private data class ConstructorParamKey(val type: TypeName, val annotations: List<AnnotationSpec>, val resolver: Resolver)
        private data class ConstructorInitializedParamKey(val type: TypeName, val initBlock: CodeBlock, val resolver: Resolver)

        override fun constructorParam(type: TypeName, annotations: List<AnnotationSpec>): String {
            return constructorParams.computeIfAbsent(ConstructorParamKey(type, annotations, resolver)) { key ->
                this.computeFieldName(key.type)
            }
        }

        override fun constructorInitialized(type: TypeName, initializer: CodeBlock): String {
            return constructorInitializedParams.computeIfAbsent(ConstructorInitializedParamKey(type, initializer, resolver)) { key ->
                this.computeFieldName(key.type)
            }
        }

        fun addFields(typeBuilder: TypeSpec.Builder) {
            constructorParams.forEach { (fd, name) ->
                typeBuilder.addProperty(
                    PropertySpec.builder(name, fd.type, KModifier.PRIVATE, KModifier.FINAL)
                        .initializer(name)
                        .build()
                )
            }
            constructorInitializedParams.forEach { (fd, name) ->
                typeBuilder.addProperty(
                    PropertySpec.builder(name, fd.type, KModifier.PRIVATE, KModifier.FINAL).build()
                )
            }
        }

        fun enrichConstructor(constructorBuilder: FunSpec.Builder) {
            constructorParams.forEach { (fd, name) ->
                constructorBuilder.addParameter(
                    ParameterSpec.builder(name, fd.type)
                        .addAnnotations(fd.annotations)
                        .build()
                )
            }
            constructorInitializedParams.forEach { (fd, name) ->
                constructorBuilder.addCode("this.%L = %L\n", name, fd.initBlock)
            }
        }

        private fun computeFieldName(type: TypeName): String {
            val qualifiedType = if (type is ParameterizedTypeName) type.rawType else type as ClassName
            val shortName = qualifiedType.simpleName.replaceFirstChar { it.lowercaseChar() }
            for (i in 1 until Int.MAX_VALUE) {
                val name = shortName + i
                if (fieldNames.add(name)) {
                    return name
                }
            }
            // never gonna happen
            throw IllegalStateException()
        }
    }

    fun applyAspects(classDeclaration: KSClassDeclaration): TypeSpec {
        val constructor = classDeclaration.findAopConstructor()
            ?: throw ProcessingErrorException("Class has no aop suitable constructor", classDeclaration)

        val typeLevelAspects: ArrayList<KoraAspect> = ArrayList()

        for (am in classDeclaration.annotations) {
            val annotationType = am.annotationType.resolveToUnderlying().declaration.qualifiedName?.asString()
                ?: continue

            for (aspect in aspects) {
                val supportedAnnotationTypes = aspect.getSupportedAnnotationTypes()
                if (supportedAnnotationTypes.contains(annotationType)) {
                    if (!typeLevelAspects.contains(aspect)) {
                        typeLevelAspects.add(aspect)
                    }
                }
            }
        }
        KoraSymbolProcessingEnv.logger.logging("Type level aspects for ${classDeclaration.qualifiedName!!.asString()}}: {$typeLevelAspects}", classDeclaration)
        val typeBuilder: TypeSpec.Builder = TypeSpec.classBuilder(classDeclaration.aopProxyName())
            .superclass(classDeclaration.toClassName())
            .addModifiers(KModifier.PUBLIC, KModifier.FINAL)
            .addAnnotation(CommonClassNames.aopProxy)

        val typeFieldFactory = TypeFieldFactory(resolver)
        val aopContext: KoraAspect.AspectContext = KoraAspect.AspectContext(typeBuilder, typeFieldFactory)

        classDeclaration.parseTags().let { tags ->
            if (tags.isNotEmpty()) {
                typeBuilder.addAnnotation(tags.makeTagAnnotationSpec())
            }
        }
        if (classDeclaration.isAnnotationPresent(CommonClassNames.root)) {
            typeBuilder.addAnnotation(CommonClassNames.root)
        }

        val classFunctions = findMethods(classDeclaration) { f ->
            !f.isConstructor() && (f.isPublic() || f.isProtected())
        }

        val methodAspectsApplied = linkedSetOf<KoraAspect>()
        classFunctions.forEach { function ->
            val methodLevelTypeAspects = typeLevelAspects.toMutableList()
            val methodLevelAspects = mutableListOf<KoraAspect>()
            val methodParameterLevelAspects = mutableListOf<KoraAspect>()
            val functionAnnotations = function.annotations.toList()
            for (am in functionAnnotations) {
                val annotationType = am.annotationType.resolveToUnderlying().declaration.qualifiedName?.asString()
                    ?: continue
                aspects.forEach { aspect ->
                    val supportedAnnotationTypes = aspect.getSupportedAnnotationTypes()
                    if (supportedAnnotationTypes.contains(annotationType)) {
                        if (!methodLevelAspects.contains(aspect)) {
                            methodLevelAspects.add(aspect)
                        }
                        methodLevelTypeAspects.remove(aspect)
                    }
                }
            }
            function.parameters.forEach { parameter ->
                for (am in parameter.annotations) {
                    val annotationType = am.annotationType.resolveToUnderlying().declaration.qualifiedName?.asString()
                        ?: continue

                    aspects.forEach { aspect ->
                        val supportedAnnotationTypes = aspect.getSupportedAnnotationTypes()
                        if (supportedAnnotationTypes.contains(annotationType)) {
                            if (!methodParameterLevelAspects.contains(aspect) && !methodLevelAspects.contains(aspect)) {
                                methodParameterLevelAspects.add(aspect)
                            }
                            methodLevelTypeAspects.remove(aspect)
                        }
                    }
                }
            }
            if (methodLevelTypeAspects.isEmpty() && methodLevelAspects.isEmpty() && methodParameterLevelAspects.isEmpty()) {
                return@forEach
            }
            KoraSymbolProcessingEnv.logger.logging(
                "Method level aspects for ${classDeclaration.qualifiedName!!.asString()}}#${function.simpleName.asString()}: {$methodLevelAspects}",
                classDeclaration
            )
            val aspectsToApply = methodLevelTypeAspects.toMutableList()
            aspectsToApply.addAll(methodLevelAspects)
            aspectsToApply.addAll(methodParameterLevelAspects)

            var superCall = "super." + function.simpleName.asString()
            val overridenMethod = FunSpec.builder(function.simpleName.asString())
                .addModifiers(KModifier.OVERRIDE)
            function.returnType?.resolve()?.let { overridenMethod.returns(it.toTypeName()) }

            if (function.modifiers.contains(Modifier.SUSPEND)) {
                overridenMethod.addModifiers(KModifier.SUSPEND)
            }

            aspectsToApply.reverse()
            val generatedMethodNames = mutableSetOf<String>()
            for (aspect in aspectsToApply) {
                val result: KoraAspect.ApplyResult = aspect.apply(function, superCall, aopContext)
                if (result is KoraAspect.ApplyResult.Noop) {
                    continue
                }

                val methodBody: KoraAspect.ApplyResult.MethodBody = result as KoraAspect.ApplyResult.MethodBody
                val baseMethodName = "_" + function.simpleName.asString() + "_AopProxy_" + aspect::class.simpleName
                var methodName = baseMethodName
                if (!generatedMethodNames.add(methodName)) {
                    for (i in 0..Int.MAX_VALUE) {
                        methodName = baseMethodName + i
                        if (generatedMethodNames.add(methodName)) {
                            break
                        }
                    }
                }

                superCall = methodName
                val f = FunSpec.builder(methodName)
                    .addModifiers(KModifier.PRIVATE)
                    .addCode(methodBody.codeBlock)

                if (function.modifiers.contains(Modifier.SUSPEND)) {
                    f.addModifiers(KModifier.SUSPEND)
                }

                function.parameters.forEach { parameter ->
                    val paramSpec = ParameterSpec.builder(parameter.name!!.asString(), parameter.type.resolve().toTypeName()).build()
                    if (!overridenMethod.parameters.contains(paramSpec)) {
                        overridenMethod.addParameter(paramSpec)
                    }
                    f.addParameter(paramSpec)
                }
                function.typeParameters.forEach { typeParameter ->
                    overridenMethod.addTypeVariable(typeParameter.toTypeVariableName())
                    f.addTypeVariable(typeParameter.toTypeVariableName())
                }
                val returnType = function.returnType!!.resolve()
                f.returns(returnType.toTypeName())
                typeBuilder.addFunction(f.build())
                methodAspectsApplied.add(aspect)
            }

            if (methodAspectsApplied.isNotEmpty()) {
                val b = CodeBlock.builder()
                if (function.returnType!!.resolve() != resolver.builtIns.unitType) {
                    b.add("return ")
                }
                b.add("%L(", superCall)
                for (i in function.parameters.indices) {
                    if (i > 0) {
                        b.add(", ")
                    }
                    val parameter = function.parameters[i]
                    b.add("%L", parameter)
                }
                b.add(")\n")
                overridenMethod.addCode(b.build())
                typeBuilder.addFunction(overridenMethod.build())
            }
        }

        val generatedClasses = mutableListOf<KClass<*>>()
        generatedClasses.add(AopSymbolProcessor::class)
        methodAspectsApplied.forEach { generatedClasses.add(it::class) }

        typeBuilder.generated(generatedClasses)

        if (classDeclaration.isAnnotationPresent(CommonClassNames.component)) {
            typeBuilder.addAnnotation(CommonClassNames.component)
        }

        val constructorBuilder = FunSpec.constructorBuilder()
        for (i in constructor.parameters.indices) {
            val parameter = constructor.parameters[i]
            typeBuilder.addSuperclassConstructorParameter("%L", parameter.name!!.asString())
            val parameterSpec = ParameterSpec.builder(parameter.name!!.asString(), parameter.type.resolve().toTypeName())

            parameter.parseTags().let { tags ->
                if (tags.isNotEmpty()) {
                    parameterSpec.addAnnotation(tags.makeTagAnnotationSpec())
                }
            }

            constructorBuilder.addParameter(parameterSpec.build())
        }
        typeFieldFactory.addFields(typeBuilder)
        typeFieldFactory.enrichConstructor(constructorBuilder)
        typeBuilder.primaryConstructor(constructorBuilder.build())
        return typeBuilder.build()
    }
}
